rfcs/20221116-enable-stochastic-rounding.md

```diff
@@ -0,0 +1,335 @@
+low-precision training can also lead to many issues like model divergence or
+quality compromises. Many studies have proven that stochastic rounding has
+substantial improvements in reducing rounding errors in low precision formats
+and helps resolve issues like divergence and quality loss. Many chips like
+Google TPU, Amazon Trainium, Tesla Dojo and Graphcore have implemented
+stochastic rounding on their hardwares. Due to the crucial role stochastic
+rounding plays in low precision training, also in order to utilize many
+hardwares’ accelerations, we propose a stochastic rounding API primarily used
+for quantized training in Tensorflow.
+
+## Background
+
+There have been many rounding schemes proposed and studied for different
+applications (IEEE 754, 2019), such as round-to-nearest, round-to-zero,
+round-to-negative-infinity, and round-to-positive-infinity, each of them having
+different round-off errors. When a large sequence of computations is
+implemented, round-off errors may be accumulated and magnified, leading to
+severe failures. Stochastic rounding differs from the standard deterministic
+modes, in that instead of always rounding to the nearest number, the decision
+about which way to round is non-deterministic and the probability of rounding up
+is proportional to the residual (value of the trailing bits that do not fit into
+the destination format). Different from deterministic rounding, stochastic
+rounding is an unbiased rounding mechanism. For example, when x = 3.7, it has a
+30% chance to be rounded to 3 and 70% chance to be rounded to 4. Thus the
+expected value of the rounding result would be 
+
+$$ 
+E(x) = 3 \times 0.3 + 4 \times 0.7 = 3.7. 
+$$
+
+Let $x ∈ R$, and let $\delta$ be the rounding precision. Then the rounded
+value $sr(x)$ of $x$ using stochastic rounding is defined as
+
+$$
+sr(x) = \begin{cases} \lfloor{x}\rfloor,& \text{with probability }
+p_1(x)=1-\frac{x-\lfloor{x}\rfloor}{\delta}\\
+\lfloor{x}\rfloor + \delta, &
+\text{with probability } p_2(x)=\frac{x-\lfloor{x}\rfloor}{\delta}\\ \end{cases}
+$$
+
+Where $\delta$ indicates the greatest representable floating/fixed-point
+number less than or equal to x.
+
+## User Benefit
+
+Users can be benefited from following aspects:
+
+-   Ready-to-use API for performing accurate stochastic rounding on many
+    different floating point formats.
+-   Performance speedups on Google TPUs.
+
+## Design Proposal
+
+During quantization, users replace high-precision floating-point numbers with
+low-precision fixed-point numbers or lower-bit floats to reduce the
+computational cost of neural network training or inference.
+
+For example, below is a code snippet that represents a part of process to
+quantize a tensor from floating-points to int8, which is then used for matrix
+multiplications:
+
+```
+inputs = tf.clip_by_value(inputs, -128.0, 127.0)
+inputs = tf.math.round(inputs) inputs = tf.cast(inputs, dtype=tf.int8)
```

Comment:
This line should probably be split into two.

Reply:
Thanks for catching! revised.